includes/api/class-wc-admin-rest-orders-controller.php

```diff
@@ -51,15 +50,59 @@ protected function prepare_objects_query( $request ) {
 
 		// Search by partial order number.
 		if ( ! empty( $request['number'] ) ) {
-			$order_ids = $wpdb->get_col(
-				$wpdb->prepare(
-					"SELECT ID FROM {$wpdb->prefix}posts WHERE post_type = 'shop_order' AND ID LIKE %s",
-					intval( $request['number'] ) . '%'
-				)
-			);
+			$partial_number = trim( $request['number'] );
+			$order_sql      = esc_sql( $args['order'] );
+			$limit          = intval( $args['posts_per_page'] );
+
+			// Handle Sequential Order Numbers.
+			if ( class_exists( 'WC_Seq_Order_Number_Pro', false ) ) {
```

Comment:
In-lieu of having conditional logic for particular extensions - don't you think this could be fixed on a per-extension basis in their own code base by hooking into REST response filters accordingly?

Reply:
I think we'd need to add some hooks to allow for extensions to easily modify the handling of the `number` parameter. _(I was going to suggest `woocommerce_rest_orders_prepare_object_query`, but that gets applied before `WC_Admin_REST_Orders_Controller` modifies `post__in`)_

Will code doing so be allowed into the Sequential Order Number extensions and are we OK with the wonky behavior for stores that aren't on the latest version of those extensions?

Reply:
I _think_ it is already filterable [here](https://github.com/woocommerce/woocommerce/blob/89740f9583893581a591651c757b6b9085cf6eb1/includes/api/v2/class-wc-rest-orders-v2-controller.php#L313) but at times I do get lost tracing the class chains in the core rest endpoints. But since all of the WC REST endpoints are extensions of core WP REST functionality, you can always use core WP filters if needed. [Here is an example](https://github.com/Automattic/wc-calypso-bridge/blob/master/store-on-wpcom/inc/wc-calypso-bridge-add-bacs-accounts.php) of me doing some _interesting_ things to manipulate some core REST responses.

While that approach I think isn't totally optimal, as all of our REST logic would execute first, it could work. But I'm fine with other approaches if we need more hooks/filters.

> Will code doing so be allowed into the Sequential Order Number extensions and are we OK with the wonky behavior for stores that aren't on the latest version of those extensions?

I think this is kind of a chicken/egg sort of scenario, but I feel like our best long-term solution for scenarios like this is to make sure there is a way for extension developers to ensure their logic can be supported within the new analytics. I just fear the moment we start adding in specific logic for extension x/y/z we will be engaging in a game of whack-a-mole.

Reply:
> I think it is already filterable [here](https://github.com/woocommerce/woocommerce/blob/89740f9583893581a591651c757b6b9085cf6eb1/includes/api/v2/class-wc-rest-orders-v2-controller.php#L313)

That's in the `prepare_object_for_response`, not `prepare_objects_query` - we're specifying a list of Order IDs to query for based on order number. The handling of the `number` parameter needs to happen before the query, not at the time of preparing the query results for the API response.

Looking at `woocommerce_rest_orders_prepare_object_query` again, I think that will work:

```php
add_filter( 'woocommerce_rest_orders_prepare_object_query', function ( $args, $request ) {
	if (
		'/wc/v4/orders' === $request['route'] &&
		isset( $request['number'] )
	) {
		// Extension handles 'number' value here and modifies $args.

		// Remove the 'number' parameter to avoid ID-based query.
		unset( $request['number'] );
	}

	return $args;
}, 10, 2 );
```

> I think this is kind of a chicken/egg sort of scenario, but I feel like our best long-term solution for scenarios like this is to make sure there is a way for extension developers to ensure their logic can be supported within the new analytics. I just fear the moment we start adding in specific logic for extension x/y/z we will be engaging in a game of whack-a-mole.

Cool. Should I open PRs for the two Sequential Order Number extensions then?

Reply:
> Cool. Should I open PRs for the two Sequential Order Number extensions then?

Sure if we can take some of the logic in the PR here and port it over I think that would be great. And would be something we can link to as an example of how to extend wc-admin via plugin logic 👍

Reply:
https://github.com/skyverge/woocommerce-sequential-order-numbers/pull/16